### ISSUE TITLE
twisted.names.client lookup functions duplicate or reference the documentation of `IResolver`

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6328]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6328|
|Type|enhancement|
|Created|2013-02-24 20:04:06Z|
|Branch|https://github.com/twisted/twisted/tree/moduleprovides-iresolver-6328-2|

In order to document the free functions in `twistd.names.client` implement most of the interface of `IResolver, so the documentation should match them. #4685 handles this by just having `@see` annotations. It would be better if we could include the documentation automatically.

See:
 * https://bugs.launchpad.net/pydoctor/+bug/1132527

"""

In [#4685](https://github.com/twisted/twisted/issues/4685)?replyto=16#comment:12 we have a series of module global functions which provide the twisted.internet.interfaces.IResolver interface.

We currently insert @see: L{twisted.internet.interfaces.IResolver.lookupAddress} so that we see links to the interface docs in the pydoctor api docs.

It would be nicer if pydoctor could find and insert the interface documentation - just like it does when it encounters an @implementer decorator.

See:
 * changeset:37229

"""

<details><summary>Searchable metadata</summary>

```
trac-id__6328 6328
type__enhancement enhancement
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_moduleprovides_iresolver_6328_2 branches/moduleprovides-iresolver-6328-2
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__names names
keywords__documentation documentation
time__1361736246000000 1361736246000000
changetime__1372372387000000 1372372387000000
version__None None
owner__rwall rwall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince set owner to @tomprince|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

[Build results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/moduleprovides-iresolver-6328)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj commented|
|-|-|

I submitted a patch to pydoctor which allows it to interpret moduleProvides.

 * https://code.launchpad.net/~richardw/pydoctor/moduleprovides-1132527/+merge/150932
 
I tested it on this branch.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj commented|
|-|-|

(In [38291]) Branching to 'moduleprovides-iresolver-6328-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Ready for review in log:branches/moduleprovides-iresolver-6328-2

mwhudson merged my pydoctor patch some time ago:
 * https://code.launchpad.net/~richardw/pydoctor/moduleprovides-1132527/+merge/150932

tomprince updated to the trunk version of pydoctor in the buildbot
documentation builder yesterday to fix [#5681](https://github.com/twisted/twisted/issues/5681).

So the API documentation should now contain full docs for the
client.lookup functions.

Summary of branch changes.
 1. I've tidied up and added some missing docstrings to IResolver.getHostByName
 2. Added a new test to verify that the t.n.client module object "provides"
    the IResolverInterface.
 3. Added client.query function to satisfy the IResolverInterface.
 4. Removed docstrings from the client.lookup functions

 * Build results: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/moduleprovides-iresolver-6328-2
 * API docs from this branch:
   https://buildbot.twistedmatrix.com/builds/apidocs/apidocs-38299.tar.bz2

I wanted to remove the docstring for
twisted.names.client.getHostByName, but that version of the function
contains an "effort" argument which is not in
IResolverSimple.getHostByName. I've raised [#6496](https://github.com/twisted/twisted/issues/6496) to suggest removing
the effort argument.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_9"></a>|@tomprince set owner to @wallrj|
|-|-|

This looks good, thanks. The only issue is that `test_query`s docstring doesn't follow the coding standard (and the docstring that it references could be improved).

Please merge after improving the docstring for `test_query`.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_10"></a>|@wallrj set status to `closed`|
|-|-|

(In [38951]) Merge moduleprovides-iresolver-6328-2

Author: rwall
Reviewer: tom.prince
Fixes: [#6328](https://github.com/twisted/twisted/issues/6328)

Use zope.interface.moduleProvides to allow pydoctor to properly
document the twisted.names.client.lookup* functions.
